### PR TITLE
feat: Initial support for custom REST authentication providers 

### PR DESCRIPTION
**Pull request by nicolatimeus**
Brief description of the PR. [e.g. Added `null` check on `object` to avoid `NullPointerException`]

* Adds API for defining custom REST authentication providers
* Added configuration options to disable built-in authentication methods
* Fixed REST requests without authentication on non-secured resources
* Added authentication provider example bundle

### CODE REVIEW

**Comment by MMaiero** [comment]
@ashmookler1 Can you please have a review?

File: kura/examples/org.eclipse.kura.example.rest.authentication.provider/META-INF/MANIFEST.MF [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+Bundle-RequiredExecutionEnvironment: JavaSE-1.8
+Service-Component: OSGI-INF/*.xml
+Bundle-ActivationPolicy: lazy
+Bundle-ClassPath: .
+Import-Package: javax.annotation;version="1.2.0",
+ javax.servlet.http;version="3.1.0",
+ javax.ws.rs.container;version="2.0.1",
+ javax.ws.rs.core;version="2.0.1",
+ org.eclipse.kura.audit;version="1.0.0",
+ org.eclipse.kura.crypto;version="1.3.0",
```
Comment (MMaiero) [inline kura/examples/org.eclipse.kura.example.rest.authentication.provider/META-INF/MANIFEST.MF]: Range needed

File: kura/examples/org.eclipse.kura.example.rest.authentication.provider/src/main/java/org/eclipse/kura/example/rest/authentication/provider/ExampleRestAuthenticationProvider.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+        final User user = (User) userAdmin.getRole(KURA_USER_PREFIX + username.get());
+
+        if ("true".equals(user.getProperties().get(KURA_NEED_PASSWORD_CHANGE))) {
+            return Optional.empty();
+        }
+
+        final String passwordHash = (String) user.getCredentials().get(KURA_PASSWORD_CREDENTIAL);
+
+        if (passwordHash == null) {
```
Comment (MMaiero) [inline kura/examples/org.eclipse.kura.example.rest.authentication.provider/src/main/java/org/eclipse/kura/example/rest/authentication/provider/ExampleRestAuthenticationProvider.java]: Objects.isNull?

**Review by MMaiero — APPROVED** [review]

**Comment by ashmookler1** [comment]
> @ashmookler1 Can you please have a review?

@MMaiero This looks good to me from a high level overview. I think this provides what is needed. Thank you!

**Review by ashmookler1 — APPROVED** [review]

**Pull request merged**